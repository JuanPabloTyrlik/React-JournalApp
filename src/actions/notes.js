import { db } from '../firebase/config';
import { loadNotes } from '../helpers/loadNotes';
import { TYPES } from '../types/types';
import Swal from 'sweetalert2';
import { fileUpload } from '../helpers/fileUpload';

export const startNewEntry = () => (dispatch, getState) => {
    const {
        auth: { uid },
    } = getState();

    const newNote = {
        title: '',
        body: '',
        imageUrl: '',
        date: new Date().getTime(),
    };

    db.collection(`${uid}/journal/notes`)
        .add(newNote)
        .then((doc) => dispatch(activeNote(doc.id, newNote)));
};

export const activeNote = (id, note) => {
    return {
        type: TYPES.NOTES_ACTIVE_ENTRY,
        payload: {
            id,
            ...note,
        },
    };
};

export const startLoadingNotes = (uid) => async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
};

export const setNotes = (notes) => {
    return {
        type: TYPES.NOTES_LOAD_ENTRIES,
        payload: notes,
    };
};

export const startSaveNote = () => (dispatch, getState) => {
    const {
        auth: { uid },
        notes: { active: note },
    } = getState();
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;
    db.doc(`${uid}/journal/notes/${note.id}`)
        .update(noteToFirestore)
        .then(() => {
            dispatch(refreshNote(note.id, noteToFirestore));
            Swal.fire('Saved', note.title, 'success');
        })
        .catch((e) => Swal.fire('Error', e.message, 'error'));
};

export const refreshNote = (id, note) => {
    return {
        type: TYPES.NOTES_UPDATE_ENTRY,
        payload: {
            id,
            ...note,
        },
    };
};

export const startUploading = (file) => async (dispatch, getState) => {
    Swal.fire({
        title: 'Uploading...',
        text: 'Please wait...',
        allowEscapeKey: false,
        allowOutsideClick: false,
        showConfirmButton: false,
        willOpen: () => {
            Swal.showLoading();
        },
    });
    const {
        notes: { active: note },
    } = getState();
    const url = await fileUpload(file);
    note.imageUrl = url;
    dispatch(refreshNote(note.id, note));
    dispatch(startSaveNote());
};
