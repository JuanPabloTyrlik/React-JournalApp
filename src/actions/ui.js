import { TYPES } from '../types/types';

export const setError = (errorMessage) => {
    return {
        type: TYPES.UI_SET_ERROR,
        payload: errorMessage,
    };
};

export const removeError = () => {
    return {
        type: TYPES.UI_REMOVE_ERROR,
        payload: null,
    };
};

export const startLoading = () => {
    return {
        type: TYPES.UI_START_LOADING,
        payload: true,
    };
};

export const finishLoading = () => {
    return {
        type: TYPES.UI_FINISH_LOADING,
        payload: false,
    };
};
