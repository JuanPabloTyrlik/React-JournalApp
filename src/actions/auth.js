import { firebase, googleAuthProvider } from '../firebase/config';
import { TYPES } from '../types/types';
import { cleanNotes } from './notes';
import { finishLoading, setError, startLoading } from './ui';

export const startLoginEmailPassword = (email, password) => (dispatch) => {
    dispatch(startLoading());
    firebase
        .auth()
        .signInWithEmailAndPassword(email, password)
        .then(({ user }) => {
            dispatch(login(user.uid, user.displayName));
        })
        .catch((e) => {
            if (e.code === 'auth/user-not-found') {
                dispatch(setError('Email is not registered'));
            } else {
                dispatch(setError(e.message));
            }
        })
        .finally(() => dispatch(finishLoading()));
};

export const startRegisterWithEmailPasswordName = (
    email,
    password,
    displayName
) => (dispatch) => {
    dispatch(startLoading());
    firebase
        .auth()
        .createUserWithEmailAndPassword(email, password)
        .then(async ({ user }) => {
            await user.updateProfile({ displayName });
            dispatch(login(user.uid, user.displayName));
        })
        .catch((e) => {
            dispatch(setError(e.message));
        })
        .finally(() => dispatch(finishLoading()));
};

export const startGoogleLogin = () => (dispatch) => {
    dispatch(startLoading());
    firebase
        .auth()
        .signInWithPopup(googleAuthProvider)
        .then(({ user: { displayName, uid } }) =>
            dispatch(login(uid, displayName))
        )
        .catch((e) => {
            dispatch(setError(e.message));
        })
        .finally(() => dispatch(finishLoading()));
};

export const login = (uid, displayName) => {
    return {
        type: TYPES.LOGIN,
        payload: {
            uid,
            displayName,
        },
    };
};

export const logout = () => {
    return {
        type: TYPES.LOGOUT,
        payload: null,
    };
};

export const startLogout = () => (dispatch) => {
    firebase
        .auth()
        .signOut()
        .then(dispatch(cleanNotes()))
        .then(dispatch(logout()));
};
