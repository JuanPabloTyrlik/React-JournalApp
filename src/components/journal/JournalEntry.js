import React from 'react';
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ id, title, body, imageUrl, date }) => {
    const noteDate = moment(date);
    const dispatch = useDispatch();

    const handleActiveNote = () => {
        dispatch(activeNote(id, { title, body, imageUrl, date }));
    };

    return (
        <div className="journal__entry pointer" onClick={handleActiveNote}>
            {imageUrl && (
                <div
                    className="journal__entry-picture"
                    style={{
                        backgroundSize: 'cover',
                        backgroundImage: `url(${imageUrl})`,
                    }}
                ></div>
            )}
            <div className="journal__entry-body">
                <p className="journal__entry-title">{title}</p>
                <p className="journal__entry-content">{body}</p>
            </div>
            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('Do')}</h4>
            </div>
        </div>
    );
};
