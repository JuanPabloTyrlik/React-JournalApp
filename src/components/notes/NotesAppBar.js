import React, { useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import moment from 'moment';
import { startSaveNote, startUploading } from '../../actions/notes';

export const NotesAppBar = () => {
    const { active: note } = useSelector((state) => state.notes);
    const dispatch = useDispatch();

    const input = useRef();

    const handleSave = () => {
        dispatch(startSaveNote());
    };

    const handleUpload = (e) => {
        input.current.click();
    };

    const handleUploadChange = (e) => {
        if (e.target.files.length > 0) {
            const file = e.target.files[0];
            dispatch(startUploading(file));
            input.current.value = '';
        }
    };

    return (
        <div className="notes__app-bar">
            <span>{moment(note.date).format('MMMM Do YYYY')}</span>
            <input
                type="file"
                name="picture"
                id="picture"
                ref={input}
                onChange={handleUploadChange}
                accept="image/*"
                style={{ display: 'none' }}
            />
            <div>
                <button className="btn" onClick={handleUpload}>
                    Picture
                </button>
                <button className="btn" onClick={handleSave}>
                    Save
                </button>
            </div>
        </div>
    );
};
