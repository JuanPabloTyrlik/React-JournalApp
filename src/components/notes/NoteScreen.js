import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { activeNote } from '../../actions/notes';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';

export const NoteScreen = () => {
    const dispatch = useDispatch();
    const { active: note } = useSelector((state) => state.notes);
    const [formValues, handleChange, reset] = useForm(note);
    const { title, body, imageUrl } = formValues;

    const activeId = useRef(note.id);
    const activeUrl = useRef(note.imageUrl);

    useEffect(() => {
        if (
            activeId.current !== note.id ||
            activeUrl.current !== note.imageUrl
        ) {
            activeId.current = note.id;
            activeUrl.current = note.imageUrl;
            reset(note);
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch(activeNote(formValues.id, { ...formValues }));
    }, [dispatch, formValues]);

    return (
        <div className="notes__main-content">
            <NotesAppBar />
            <div className="notes__content">
                <input
                    type="text"
                    name="title"
                    id="title"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    value={title}
                    onChange={handleChange}
                />
                <textarea
                    name="body"
                    id="text-area"
                    cols="20"
                    rows="5"
                    placeholder="What happened today?"
                    className="notes__text-area"
                    value={body}
                    onChange={handleChange}
                ></textarea>
                {formValues.imageUrl && (
                    <div className="notes__image">
                        <img src={imageUrl} alt="image" />
                    </div>
                )}
            </div>
        </div>
    );
};
