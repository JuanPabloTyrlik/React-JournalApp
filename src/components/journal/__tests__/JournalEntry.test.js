import React from 'react';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { JournalEntry } from '../JournalEntry';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const store = mockStore({
    ui: {
        loading: false,
        msgError: null,
    },
    auth: {},
    notes: {
        notes: [
            {
                id: 'kOs0NIVIeAeDQL5xEWg4',
                imageUrl:
                    'https://res.cloudinary.com/dxd81c3e7/image/upload/v1606084697/h2iks4zr4bjcxtj13k6g.png',
                body: 'Dummy Text',
                date: 1605738681205,
                title: 'New Note',
            },
            {
                id: 'MSe9uD5X6lcVEBG82YnW',
                date: 1605650153830,
                body: 'lorem ipsium',
                title: 'Nota 11',
                imageUrl:
                    'https://res.cloudinary.com/dxd81c3e7/image/upload/v1606084704/wxdq8dbsr6wii77ojwz7.webp',
            },
        ],
        active: {
            id: 'kOs0NIVIeAeDQL5xEWg4',
            title: 'New Note',
            body: 'Dummy Text',
            imageUrl:
                'https://res.cloudinary.com/dxd81c3e7/image/upload/v1606084697/h2iks4zr4bjcxtj13k6g.png',
            date: 1605738681205,
        },
    },
});

store.dispatch = jest.fn();

jest.mock('../../../actions/notes', () => ({
    activeNote: jest.fn(),
}));

describe('Tests on JournalEntry', () => {
    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...store.getState().notes.active} />
        </Provider>
    );

    beforeEach(() => {
        jest.clearAllMocks();
        store.clearActions();
    });
    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test('should set active note', () => {
        wrapper.find('.journal__entry').simulate('click');
        expect(activeNote).toHaveBeenLastCalledWith('kOs0NIVIeAeDQL5xEWg4', {
            body: 'Dummy Text',
            date: 1605738681205,
            imageUrl:
                'https://res.cloudinary.com/dxd81c3e7/image/upload/v1606084697/h2iks4zr4bjcxtj13k6g.png',
            title: 'New Note',
        });
    });
});
