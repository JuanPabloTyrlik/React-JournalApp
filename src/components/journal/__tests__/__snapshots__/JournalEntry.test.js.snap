// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on JournalEntry should render correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://res.cloudinary.com/dxd81c3e7/image/upload/v1606084697/h2iks4zr4bjcxtj13k6g.png)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      New Note
    </p>
    <p
      className="journal__entry-content"
    >
      Dummy Text
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      18th
    </h4>
  </div>
</div>
`;
